feat: generate TypeScript types on save

lihbr opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

As requested by @Duaner, this PR adds an early integration of TypeScript inside Slice Machine by generating types for Slices and Custom Types on save and rename.

- A `types.ts` file is now created alongside Custom Types `index.json` model;
- A `types.ts` file is now created alongside Slices `model.json` model.

⚠ &nbsp;This code will have to be removed when switching to plugins. Plugins are already handling TypeScript integration the same way as the one implemented in this PR.

<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it resolves an open issue, please link to the issue here. For example "Resolves: #137" -->

## Checklist:

<!--- Put an `x` in all the boxes that apply. -->
<!--- If your change requires a documentation PR, please link it appropriately -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] All new and existing tests are passing. (existing tests were failing on Windows because of win32/posix paths inconsistencies)

<!--- A cute animal picture is welcome to close your PR! -->

lihbr committed:
feat: generate TypeScript types on save

Heyrows commented on packages/slice-machine/server/src/api/io/CustomType.ts:
```diff
@@ -14,8 +15,21 @@ export function writeCustomType(path: string, customType: CustomTypeSM) {
   customType.label = newCustomTypeName;
-  writeCustomType(path, customType);
+  writeCustomType(modelPath, customType);
+  writeCustomTypeTypes(typesPath, customType);
```
nothing changes on the types, why do we need to re-write them ?

lihbr replied:
TSDocs on the generated type make use of the Slice Display name.

Heyrows commented on packages/slice-machine/server/src/api/io/Slice.ts:
```diff
@@ -11,8 +12,19 @@ export function writeSlice(path: string, slice: SliceSM) {
   slice.name = newSliceName;
-  writeSlice(path, slice);
+  writeSlice(modelPath, slice);
+  writeSliceTypes(typesPath, slice);
```
same question as the rename custom types, why do we need to rebuild the types if we rename them ?

lihbr replied:
TSDocs on the generated type make use of the Slice Display name here too ☺️

MarcMcIntosh commented on packages/core/src/node-utils/paths.ts:
```diff
@@ -35,6 +37,7 @@ const Paths = (cwd: string, prefix: string): Paths => ({
   customType: (id: string) => ({
     value: () => path.join(Paths(cwd, prefix).value()),
     model: () => path.join(Paths(cwd, prefix).value(), id, "index.json"),
+    types: () => path.join(Paths(cwd, prefix).value(), id, "types.ts"),
```
isn't the standard to call it `index.d.ts` ?

lihbr replied:
As far as I know, `*.d.ts` files are used to provide type information about their eponymous JavaScript files (e.g. `createFoo.js` & `createFoo.d.ts`)

In this case, `types.ts` doesn't represent the type of the `index.js` component (or the exports from that file). It represents the types of the Slice object (the `index.js` component eventually uses that). With that in mind we went for `types.ts` as thought more appropriate.

MarcMcIntosh replied:
ok

Heyrows reviewed (commented):
Should Slice template be modified to include the types ?

Thank you for contributing to Slice Machine 😁.
As you're adding a new feature, could you add the corresponding test / modify the existing one to match the expected behavior please 🙏

MarcMcIntosh reviewed (commented):
How would you test this and can it be tested?

lihbr committed:
chore: update slice templates

lihbr commented:
> Should Slice template be modified to include the types ?

Done, for now, it's only possible to streamline it for templates using `@prismicio/react`.

`prismic-dom`, `prismic-reactjs`, and `@prismicio/vue@2` cannot have it automated the same way because the component prop types depend on the user setup/the slice zone they use. Users, especially TypeScript ones, will still be able to take advantage of it, but manually, not from the template.

> As you're adding a new feature, could you add the corresponding test / modify the existing one to match the expected behavior please 🙏
>
> How would you test this and can it be tested?

`prismic-ts-codegen` is the library doing the bulk of the work and is already [extensively tested](https://app.codecov.io/gh/prismicio/prismic-ts-codegen). Regarding its integration with Slice Machine, we'd be mainly interested in testing that the types file is indeed created and updated when needed.

In that regard, I haven't seen tests regarding Slices/Custom Types save events (unit & e2e). If I missed them let me know where I can find them, happy to add a few. If this part of the codebase is not covered yet, I don't think it's part of the scope of this PR to start covering it (especially since this PR's additions are temporary).

lihbr committed:
test(e2e): add assertions for `types.ts` files

lihbr committed:
style: accomodate prettier

MarcMcIntosh commented on cypress/integration/slices/00-create.spec.js:
```diff
@@ -35,6 +35,7 @@ describe("Create Slices", () => {
       "eq",
       `/${lib}/${sliceName}/default`
     );
+    cy.readFile(`${path}/${sliceName}/types.ts`).should("contains", sliceName);
```
nice

MarcMcIntosh commented on cypress/integration/customTypes/00-create.spec.js:
```diff
@@ -34,5 +36,9 @@ describe("Custom Types specs", () => {
     cy.get('[data-cy="custom-type-secondary-breadcrumb"]').contains(
       `/ ${name} - Edited`
     );
+    cy.readFile(`${path}/${id}/types.ts`).should(
```
nice

MarcMcIntosh commented on cypress/integration/slices/00-create.spec.js:
```diff
@@ -50,5 +51,9 @@ describe("Create Slices", () => {
     cy.get('[data-cy="slice-and-variation-name-header"]').contains(
       `/ ${editedSliceName} / Default`
     );
+    cy.readFile(`${path}/${editedSliceName}/types.ts`).should(
```
nice

MarcMcIntosh commented on packages/slice-machine/server/src/api/slices/create/index.ts:
```diff
@@ -35,6 +35,7 @@ const copyTemplate = async (
     await copy(templatePath, path.join(env.cwd, from, sliceName), {
       componentName: sliceName,
       componentId: snakelize(sliceName),
+      pascalId: pascalize(sliceName),
```
what's this for?

Heyrows replied:
wait does that mean if it's rename we should change that part as well ?

lihbr replied:
Types are generated after the pascalcased ID by convention.

Upon a Slice rename the ID wouldn't change so the type definition would remain the same and nothing would break here :)

chengoconnell committed:
feature(): dont fetch remote cts if local ct folder is empty

chengoconnell committed:
feature(): add changes button to sidebar

chengoconnell committed:
Merge pull request #534 from prismicio/aoc/724-dont-pull-remote-ct-and-slices

Aoc/724 dont pull remote ct and slices

chengoconnell committed:
feature(): add push changes button and changes page

chengoconnell committed:
feature(): add changes indicator and pass in number of changes prop

chengoconnell committed:
refactor(): abstract out titles to own components and add future todos

chengoconnell committed:
refactor(): remove unused import

chengoconnell committed:
refactor(): make link object more generic

chengoconnell committed:
feature(): use theme ui colors

chengoconnell committed:
Merge pull request #537 from prismicio/aoc/728-new-changes-page-empty

Aoc/728 new changes page empty

chengoconnell committed:
feature(): add status badge with tooltip

chengoconnell committed:
feature(): add column to custom type page

chengoconnell committed:
feature(): add fullstop

chengoconnell committed:
wip

chengoconnell committed:
refactor(): remove import and make function consistent

chengoconnell committed:
feature(): make tooltip stick and apply max width

MarcMcIntosh committed:
Merge pull request #511 from prismicio/v0.4.2

[Release Branch] v0.4.2

chengoconnell committed:
feature(): compensate for added padding by reducing margin

MarcMcIntosh committed:
Publish

 - @slicemachine/client@1.1.12
 - @slicemachine/core@1.1.12
 - @slicemachine/init@1.1.12
 - slice-machine-ui@0.4.2

MarcMcIntosh committed:
deps: update pupiteer

MarcMcIntosh committed:
ci: ensure alpha preid is used

MarcMcIntosh committed:
Publish

 - slice-machine-ui@0.4.3-alpha.0

MarcMcIntosh committed:
Publish

 - slice-machine-ui@0.4.3-alpha.1

VarunBalakrishna committed:
fix: move `checkSimulatorSetup` to when taking screenshot with simulator rather than when uploading custom screenshot file

chengoconnell committed:
Merge pull request #548 from prismicio/aoc/725-custom-types-status-label

Aoc/725 custom types status label

Commit:
add slice machine init before cypress tests

Commit:
re-add the cypress retries

Commit:
fix cypress test

Commit:
use cypress account

Commit:
fix PR comments

Commit:
change the repo name to be lowercase

chengoconnell committed:
feature(): update customtypestatus function

chengoconnell committed:
feature(): update backend to compute status

chengoconnell committed:
feature(): dynamically render status and update slicesm type

chengoconnell committed:
feature(): update redux status for save create and push

chengoconnell committed:
feature(): allow button to take in status

Commit:
add unsyced slices

Commit:
change section block header

Commit:
use secrets email and password

chengoconnell committed:
feature(): update status badge and custom type function

chengoconnell committed:
refactor(): move where we initialise the status

Commit:
change position of e2e test directory

MarcMcIntosh committed:
Merge branch 'v0.4.3' into next-release

chengoconnell committed:
feature(): make sure available custom types and selected custom type gets updated states

MarcMcIntosh committed:
Merge pull request #556 from prismicio/next-release

Next release

MarcMcIntosh committed:
Publish

 - slice-machine-ui@0.4.3-alpha.2

Commit:
fix PR comments

chengoconnell committed:
feature(): update state for available custom types when we push and save

sintinilorenzo committed:
Merge pull request #555 from prismicio/ls/729-add-unsynced-slices-in-changes-page

add unsyced slices

chengoconnell committed:
feature(): change import

sintinilorenzo committed:
Publish

 - slice-machine-ui@0.4.3-alpha.3

chengoconnell committed:
feature(): pass in entire custom type instead of id

chengoconnell committed:
feature(): handle unknown case when we don't get any remote custom types

Commit:
set wroom.io as the cypress url for pipelines

chengoconnell committed:
Merge pull request #553 from prismicio/aoc/740-custom-types-status-with-logic

Aoc/740 custom types status with logic

Commit:
add the CYPRESS_URL within the secrets

chengoconnell committed:
Publish

 - @slicemachine/client@1.1.13-alpha.0
 - @slicemachine/core@1.1.13-alpha.0
 - @slicemachine/init@1.1.13-alpha.0
 - slice-machine-ui@0.4.3-alpha.4

sintinilorenzo committed:
Merge pull request #552 from prismicio/ls/744-add-cypress-test-to-initialise-slice-machine

add bash script to test the slice machine ini

sintinilorenzo committed:
Publish

 - @slicemachine/init@1.1.13-alpha.1
 - slice-machine-ui@0.4.3-alpha.5

hypervillain committed:
set default dns result to ipv4 (node17,18)

Commit:
fix dev server scripts by using rollup

Commit:
update the status on the available custom types considering the remote CTs

Commit:
move method

sintinilorenzo committed:
Merge pull request #560 from prismicio/ls/740-fix-dev-server-scripts

Ls/740 fix dev server scripts

sintinilorenzo committed:
Publish

 - slice-machine-ui@0.4.3-alpha.6

Commit:
add cts-to-changes-page

Commit:
fix CR comments

sintinilorenzo committed:
Merge pull request #561 from prismicio/ls/730-add-cts-to-changes-page

add cts-to-changes-page

sintinilorenzo committed:
Publish

 - slice-machine-ui@0.4.3-alpha.7

Commit:
show in the changes indicator the number of unsynced slices/cts

sintinilorenzo committed:
Merge pull request #562 from prismicio/ls/731-display-changes-in-navitem

show in the changes indicator the number of unsynced slices/cts

sintinilorenzo committed:
Publish

 - slice-machine-ui@0.4.3-alpha.8

Commit:
add an empty page when there are no changes to push

sintinilorenzo committed:
Merge pull request #563 from prismicio/ls/732-add-empty-page-for-no-changes

add an empty page when there are no changes to push

sintinilorenzo committed:
Publish

 - slice-machine-ui@0.4.3-alpha.9

Commit:
add offline page

Commit:
remove unused variable

sintinilorenzo committed:
Merge pull request #564 from prismicio/ls/733-add-offline-page

add offline page

sintinilorenzo committed:
Publish

 - slice-machine-ui@0.4.3-alpha.10

Commit:
add unauthorised page

Commit:
open the login modal on button click

Commit:
make the push button disabled for unauhthrised statuses

chengoconnell committed:
wip

Commit:
make parameter optional

Commit:
add individual spinner to each slice

Commit:
fix PR comments

sintinilorenzo committed:
Merge pull request #566 from prismicio/aoc/735-unsynced-slices-take-screenshot

Aoc/735 unsynced slices take screenshot

sintinilorenzo committed:
Publish

 - slice-machine-ui@0.4.3-alpha.11

hypervillain committed:
replace node-fetch with axios

Commit:
add pushSaga

MarcMcIntosh committed:
core: add function to read a file and generate a hash

MarcMcIntosh committed:
client: add file hash in to sccreenshot url rather than uniqid

Heyrows committed:
fix push reducer on Custom type list

MarcMcIntosh committed:
core: add hash to screenshot

Heyrows committed:
add test for push action on availableCustomType reducer

Heyrows committed:
Merge pull request #568 from prismicio/ac/fix-push-state

Fix Custom Type status after push action.

Heyrows committed:
Publish

 - slice-machine-ui@0.4.3-alpha.12

Heyrows committed:
hide number of changes if the user doesn't have internet

MarcMcIntosh committed:
sm-ui: compare imageUrl paths to check if screenshot is up to date

Heyrows committed:
unify slice card description

chengoconnell committed:
feature(): add ct status when user is offline

chengoconnell committed:
feature(): set status to unknown disconnected if user is not logged in on the backend

Commit:
add scenario 001

Commit:
add datacu prop to other zones

Heyrows committed:
review changes

MarcMcIntosh committed:
sm-ui: save imageUrl before pushing slice

Commit:
fix typo

MarcMcIntosh committed:
sm-ui: fix type

MarcMcIntosh committed:
refactor: screenshot hash, remove un-needed hash property

chengoconnell committed:
feature(): when user is disconnected make sure ct status stays the same

Commit:
add comment for the files called from the setup

sintinilorenzo committed:
Merge pull request #571 from prismicio/ls/767-add-scenario-001

add scenario 001

chengoconnell committed:
feature(): remove unknown default case

chengoconnell committed:
feature(): create new object instead of modifying existing one

chengoconnell committed:
feature(): show push to prismic button if status is anything but synced

sintinilorenzo committed:
Publish

 - @slicemachine/init@1.1.13-alpha.2
 - slice-machine-ui@0.4.3-alpha.13

chengoconnell committed:
feature(): undo uneccessary check for rename custom type

MarcMcIntosh committed:
Merge pull request #573 from prismicio/mm/screenshot-sync

Mm/screenshot sync

MarcMcIntosh committed:
Publish

 - @slicemachine/client@1.1.13-alpha.1
 - @slicemachine/core@1.1.13-alpha.1
 - @slicemachine/init@1.1.13-alpha.3
 - slice-machine-ui@0.4.3-alpha.14

Heyrows committed:
Merge pull request #569 from prismicio/ac/unify-design

Unify slice card design.

Heyrows committed:
Publish

 - slice-machine-ui@0.4.3-alpha.15

MarcMcIntosh committed:
changes: handle error responses

MarcMcIntosh committed:
sm-ui(changes): refresh state after pushing

MarcMcIntosh committed:
sm-ui(changes): filter out synced slices

MarcMcIntosh committed:
Merge branch 'v0.4.3' into ls/736-push-changes

MarcMcIntosh committed:
sm-ui(changes): filter out synced slices when pushing

Heyrows committed:
Merge pull request #570 from prismicio/aoc/740-status-with-logic-fix

feature(): add ct status when user is offline

Heyrows committed:
Merge pull request #565 from prismicio/ls/734-add-auth-error-page

Ls/734 add auth error page

Heyrows committed:
Publish

 - slice-machine-ui@0.4.3-alpha.16

Heyrows committed:
Merge pull request #559 from prismicio/fix-node-dns-node-detch

set default dns result to ipv4 (node17,18)

MarcMcIntosh committed:
Merge branch 'v0.4.3' into ls/736-push-changes

Heyrows committed:
Merge branch 'v0.4.3' into bug/fix-custom-screenshot-checking-simulator

Heyrows committed:
Merge pull request #576 from prismicio/VarunBalakrishna/fix-custom-screenshot-checking-simulator

[Fix] Check simulator steps for regular screenshots not custom

Heyrows committed:
Publish

 - slice-machine-ui@0.4.3-alpha.17

MarcMcIntosh committed:
sm-ui(changes): do them sequentually

Heyrows committed:
add try catch to tracking events.

Heyrows committed:
Merge branch 'v0.4.3' into ac/non-blocking-tracking

Heyrows committed:
Merge pull request #577 from prismicio/ac/non-blocking-tracking

[Fix] Non blocking events.

Heyrows committed:
Publish

 - slice-machine-ui@0.4.3-alpha.18

MarcMcIntosh committed:
sm-ui(pushSaga): add test cases

MarcMcIntosh committed:
client: use non zero exit code when using process.exit

MarcMcIntosh committed:
test(sm-ui): update getEnv test

MarcMcIntosh committed:
sm-ui(getEnv): use non zero value when calling process.exit

Heyrows committed:
fix tests

MarcMcIntosh committed:
chore: add esbuild to slice-machine-ui dev deps

MarcMcIntosh committed:
Merge pull request #578 from prismicio/mm/790-push-sequentually

sm-ui(changes): do them sequentually

Heyrows committed:
Merge pull request #579 from prismicio/mm/fix-test-exit-code

client: use non zero exit code when using process.exit

MarcMcIntosh committed:
Merge branch 'v0.4.3' into ls/736-push-changes

Heyrows committed:
Publish

 - @slicemachine/client@1.1.13-alpha.2
 - @slicemachine/init@1.1.13-alpha.4
 - slice-machine-ui@0.4.3-alpha.19

MarcMcIntosh committed:
sm-ui(changes): push the slices in parallel then the custom-types.

MarcMcIntosh commented:
Update the branch :)

MarcMcIntosh committed:
sm-ui:(changes): check that pushSliceCreator.failure is called

MarcMcIntosh committed:
sm-ui(changes): pass handleError function to pushchangesSaga

MarcMcIntosh committed:
sm-ui(changes): set errors to null when sucessfull

MarcMcIntosh committed:
sm-ui(changes): remove unneeded type sugnature

MarcMcIntosh committed:
sm-ui(changes): refresh state if everything goes well

MarcMcIntosh committed:
Revert "sm-ui(changes): refresh state if everything goes well"

This reverts commit 6fc91bb95f2fe7acd4dfc0fea47395246deb8775.

MarcMcIntosh committed:
sm-ui(changes): remove un-needed useEffect

MarcMcIntosh committed:
sm-ui(changes): remove un-used import

MarcMcIntosh committed:
sm-ui(changes): remove comments

MarcMcIntosh committed:
Merge pull request #575 from prismicio/ls/736-push-changes

[WIP]: Ls/736 push changes

MarcMcIntosh committed:
Publish

 - slice-machine-ui@0.4.3-alpha.20

lihbr committed:
Merge branch 'v0.4.3' of github.com:prismicio/slice-machine into lh/typescript-integration

lihbr committed:
refactor(deps): update `prismic-ts-codegen` integration

lihbr commented:
Closed in favor of #583
